### PR TITLE
[Messenger] Add some UPGRADE entries regarding 4.2 BC breaks

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 4.2 
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | N/A   
| License       | MIT
| Doc PR        | N/A

Just highlighting most relevant changes for users and their upgrade paths. 
For exhaustivity, you'll still have to read the Messenger CHANGELOG file.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Does it account for #29010 and maybe #29006 already?

File: UPGRADE-4.2.md [hunk 1]
```diff
@@ -125,6 +125,26 @@ HttpKernel
...
+
+   After:
+   ```yaml
+   framework:
+       messenger:
+           events:
+               default_middleware: allow_no_handlers
+   ```
+
+ * The `messenger:consume-messages` command expects a mandatory `--bus` option value if you have more than one bus configured.
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: doesn't it consume message for all buses when none is provided?

File: UPGRADE-4.2.md [hunk 2]
```diff
@@ -161,8 +181,66 @@ FrameworkBundle
 Messenger
 ---------
 
- * `EnvelopeItemInterface` doesn't extend `Serializable` anymore
- * The `handle` method of the `Symfony\Component\Messenger\Middleware\ValidationMiddleware` and `Symfony\Component\Messenger\Asynchronous\Middleware\SendMessageMiddleware` middlewares now requires an `Envelope` object to be given (because they implement the `EnvelopeAwareInterface`). When using these middleware with the provided `MessageBus`, you will not have to do anything. If you use the middlewares any other way, you can use `Envelope::wrap($message)` to create an envelope for your message.
+ * The `MiddlewareInterface::handle()` and `SenderInterface::send()` methods must now return an `Envelope` instance.
+ * Handlers must not return any value (`void`). If you used to return a value, e.g in query bus handlers, you can either:
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: The return value of handlers is ignored. *etc.*

File: UPGRADE-4.2.md [hunk 3]
```diff
@@ -161,8 +181,66 @@ FrameworkBundle
...
+      // When dispatching:
+      $bus->dispatch(new Query([$this, 'onResult']));
+
+      // In your handler:
+      $query->executeCallback($yourResult);
+      ```
+
+ * The `EnvelopeAwareInterface` was removed and the `MiddlewareInterface::handle()` method now requires an `Envelope` object
+   as first argument. When using built-in middleware with the provided `MessageBus`, you will not have to do anything.  
+   If you use your own `MessageBusInterface` implementation, you can use `Envelope::wrap($message)` to create an envelope for your messages.
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: `Envelope::wrap()` doesn't exist anymore

File: UPGRADE-4.2.md [hunk 4]
```diff
@@ -161,8 +181,66 @@ FrameworkBundle
...
+       return $envelope;
+   }
+   ```
+ * `StampInterface` replaces `EnvelopeItemInterface` and doesn't extend `Serializable` anymore.
+    Built-in `ReceivedMessage`, `ValidationConfiguration` and `SerializerConfiguration` where renamed
+    respectively `ReceivedStamp`, `ValidationStamp`, `SerializerStamp` and moved in the `Stamp` namespace.
+ * `AllowNoHandlerMiddleware` has been removed in favor of a new constructor argument on `HandleMessageMiddleware`
+ * The `ConsumeMessagesCommand` class now takes an instance of `Psr\Container\ContainerInterface`
+    as first constructor argument, i.e a message bus locator. The CLI command will now expects a mandatory 
+    `--bus` option value if there is more than one bus in the locator.
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: see question above about `--bus|`

**Review by nicolas-grekas — COMMENTED** [review]
nice thanks

File: UPGRADE-4.2.md (hunk 1 above)
Comment (ogizanagi) [reply UPGRADE-4.2.md]: Nope. The `--bus` option is about where the received message is dispatched back. It's independent from where was initially dispatched the message.
If none is provided, interactively, we'll ask to select one.

File: UPGRADE-4.2.md (hunk 3 above)
Comment (ogizanagi) [reply UPGRADE-4.2.md]: Right. In my mind it was re-introduced in another PR but was closed instead 😅

**Comment by ogizanagi** [comment]
Nope. I'm not sure those deserve more detailed entries. This is more about the outer-boundaries of the component usage & config. Unless you have more suggestions :)

**Review by nicolas-grekas — APPROVED** [review]

File: UPGRADE-4.2.md [hunk 5]
```diff
@@ -125,6 +125,26 @@ HttpKernel
 FrameworkBundle
 ---------------
 
+ * The `allow_no_handler` middleware has been removed. Use `framework.messenger.[bus].default_middleware` instead:
```
Comment (OskarStark) [inline UPGRADE-4.2.md]: You say the handler is removed, but below is „allow_no_handlers“... it is a bit confusing

Comment (ogizanagi) [reply UPGRADE-4.2.md]: Actually the **middleware** is removed in favor of a flag in the `HandleMessageMiddleware`. I believe this information isn't really useful from the config POV so I omit it and just talk about setting the `framework.messenger.[bus].default_middleware` config option to `allow_no_handlers` to set this flag.
How can I improve this?

Comment (ro0NL) [reply UPGRADE-4.2.md]: if it's too hard to explain.. i suggested https://github.com/symfony/symfony/pull/28945#discussion_r227114703 :angel:

Comment (ogizanagi) [reply UPGRADE-4.2.md]: @ro0NL : At least, I agree`allow_no_handlers` at the bus node level would be self-explanatory.

Comment (ro0NL) [reply UPGRADE-4.2.md]: it also brings more flexibility, i.e. keep logging but disable other defaults.

Being able to control it per case, IMHO would be more straightforward. Currently it's a bit all or nothing, and you need to know what default_middleware=true actually means.

Comment (ogizanagi) [reply UPGRADE-4.2.md]: That's why I was never really convinced by the default middleware config over explicit middleware stack configuration.

Comment (ro0NL) [reply UPGRADE-4.2.md]: but then we cant provide sensible defaults (including default ordering).. so having a node per default + custom middleware stack seems like best of both.

So the only "weird" case is, if you want re-order defaults

```yaml
bus:
  logging: false # log before
  middleware: [custom, logging] # log after
```

Comment (ogizanagi) [reply UPGRADE-4.2.md]: That's still requires knowledge about what the defaults are and their order.

We could add a `framework.messenger.default_middleware` with `before` and `after` entries (or a `[middleware]` placeholder for specific bus middleware), with same format as middlewares & default values set in the `Configuration` allowing these to be dumped on `debug:config` and changed in userland so anyone can provide their own defaults, but not sure it is worth.

```yml
framework:
    messenger:
        default_middleware: ~
        # Which by default is the same as:
        # default_middleware:
        #   before:
        #     - logging
        #   after:
        #     - route_messages
        #     - call_message_handler
        bus:
            command:
                middleware:
                    - doctrine_transaction
            query:
                default_middleware: false
                middleware:
                    - cache
                    - call_message_handler
            events:
                allow_no_handlers: true
```

But defaults may also be a PITA if one day we need to deprecate a middleware part of defaults in favor of another one. We cannot either add more or remove one in the future.

IMHO, explicit is always better.

**Review by sroze — APPROVED** [review]
Fabulous, thank you for taking care of this ❤️

File: UPGRADE-4.2.md (hunk 5 above)
Comment (ro0NL) [reply UPGRADE-4.2.md]: maybe you're right :+1: the defaults are not super complex to declare anyway...

File: UPGRADE-4.2.md [hunk 6]
```diff
@@ -161,8 +183,66 @@ FrameworkBundle
...
+   {
+       // do something before
+       $envelope = $stack->next()->handle($envelope, $stack);
+       // do something after
+    
+       return $envelope;
+   }
+   ```
+ * `StampInterface` replaces `EnvelopeItemInterface` and doesn't extend `Serializable` anymore.
+    Built-in `ReceivedMessage`, `ValidationConfiguration` and `SerializerConfiguration` where renamed
```
Comment (chalasr) [inline UPGRADE-4.2.md]: s/where/were (or have been?)

File: UPGRADE-4.2.md [hunk 7]
```diff
@@ -161,8 +183,66 @@ FrameworkBundle
...
+    respectively `ReceivedStamp`, `ValidationStamp`, `SerializerStamp` and moved in the `Stamp` namespace.
```
Comment (chalasr) [inline UPGRADE-4.2.md]: should be `moved into` or `moved to` I believe

File: UPGRADE-4.2.md [hunk 8]
```diff
@@ -161,8 +183,66 @@ FrameworkBundle
...
+ * `AllowNoHandlerMiddleware` has been removed in favor of a new constructor argument on `HandleMessageMiddleware`
+ * The `ConsumeMessagesCommand` class now takes an instance of `Psr\Container\ContainerInterface`
+    as first constructor argument, i.e a message bus locator. The CLI command will now expects a mandatory
```
Comment (chalasr) [inline UPGRADE-4.2.md]: I would remove `will`. If you disagree, `expects` should be `expect` :)

**Review by chalasr — APPROVED** [review]
Much appreciated!

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @ogizanagi.

**Pull request merged**